### ISSUE TITLE
Monad Transformers for `smash-core`

### ISSUE THREAD

**Issue by emilypi** [emilypi/smash#7]
`transformers` and `mtl` are bundled with GHC, so this is a zero-cost addition to the library.

**Comment by darcykimball** [comment]
I'm not sure I can help, but I'm curious: for `transformers`, does this mean something like adding a module named `Control.Monad.Trans.Smash` with stuff like:

```haskell
newtype SmashT m a b = SmashT { runSmashT :: m (Smash a b) }

instance Functor m => Functor (SmashT m a) where
  fmap f = SmashT . fmap (fmap f) . runSmashT

instance (Monoid a, Applicative m) => Applicative (SmashT m a) where
  pure = SmashT . pure . Smash mempty
  f <*> x =
    let fsm = runSmashT f
        xsm = runSmashT x
     in SmashT $ (<*>) <$> fsm <*> xsm

instance (Monoid a, Monad m) => Monad (SmashT m a) where
  x >>= f = SmashT $ do
    sm <- runSmashT x
    case sm of
      Nada -> return Nada
      Smash a b -> do
        sm' <- runSmashT (f b)
        case sm' of
          Nada -> return Nada
          Smash a' b' -> return $ Smash (a `mappend` a') b'
```
...plus all the usual instances for `MonadTrans`, etc.?

Also, for `mtl`, would `Smash`, for example, have an instance for `MonadError`? Would there be any others? I'm sorta lacking an intuition for what the types in this library can model: it seems to me like `Smash a b` can be used like `WriterT a (MaybeT Identity) b` as a writer that may fail; would this be a usual usage?

**Comment by emilypi** [comment]
```
does this mean something like adding a module...
```

Yes! that's all it means. 

```

**Comment by darcykimball** [comment]
I'll give a try at helping out. What nixpkgs/nixos version(s) are you guys using to test?

**Comment by emilypi** [comment]
I'm not using any nix infra to test any of these packages - it's purely travis. this is because I'm not a Nix person, and I use a very dumb tool called `cabal2nix` to nixify my packages for the nix folks. If you have a better or more competent way, I am all ears 😄

**Comment by emilypi** [comment]
Yeah, so just to follow up on this, the implementations are a little bit uninteresting, but otherwise, don't look terrible: 

```
newtype CanT m a b = CanT  { runCanT :: m (Can a b) } 

instance (w ~ (), MonadWriter w m) => MonadWriter w (CanT m w) where
  tell () = pure ()
  listen (CanT m) = CanT (go <$> m) where
    go = \case
      Non -> Non
      One () -> One ()
      Eno b -> Eno (b,())
      Two () b -> Two () (b, ())

  pass (CanT m) = CanT (go <$> m) where
    go = \case
      Non -> Non
      One () -> One ()
      Eno (a, _) -> Eno a
      Two a (b, w) -> Two (w a) b

```

**Comment by emilypi** [comment]
Closed via #25
